### DISCUSSION TITLE
Add commands for number-based buffer switching

### DISCUSSION THREAD

**Discussion by tmke8** [helix-editor/helix#10353] [Ideas]
I know I can switch buffers with space+b and then searching or moving up-and-down, but this still introduces some kind of mental cost for buffer switching, especially when you do it frequently. I also feel like a _modern_ editor should show a list of the open files.

With (n)vim, both of these problems are solved by [buftabline.nvim](https://github.com/jose-elias-alvarez/buftabline.nvim). It shows the open buffers in the "tabline" and allows switching to them with <leader>+buffer number. This allows fast access to the first 10 buffers, at almost no mental cost. (From experience I can tell you that once you get used to this, all other methods of buffer switching will feel excruciatingly slow.)

I know that keymappings in helix deliberately do not make use of the numbers, because they're farther away, but this isn't really an editing command, so I think it's fine in this case, though I don't know how you see this.

**Comment by pickfire** [comment]
Personally I don't like this because it waste space (I am on a 80x24 terminal). May seemed useful at times but I might still use `space b` for now. If it shows up only when `space` is pressed then I am okay with it.

**Comment by kirawi** [comment]
Could be something configurable.

**Comment by antoyo** [comment]
As an additional argument, I find this feature almost required to make any use of the commands `goto_next_buffer` and `goto_previous_buffer` as we currently have no way to know what are the next and previous buffers.

**Comment by kirawi** [comment]
As a bufferline implementation was merged a while ago, I'm marking this as a `good-first-issue` for the fast buffer switching shortcut.

**Comment by Elias-Graf** [comment]
I have implemented the command `buffer-goto` here: https://github.com/helix-editor/helix/pull/4756. 

I've currently implemented it only as a normal, typed command. So no `+<i>`, nor <kbd>space</kbd> + <kbd>&lt;i&gt;</kbd>. 

Usage: `:buffer-goto <i>` or `:b <i>`. 

I've also refactored the "previous" / "next" buffer commands a bit, as there was some functional overlap, and a naming collision with the function `goto_buffer(&Editor, Direction)` (now named `goto_buffer_by_direction`).

I personally would love to see switching buffers using a shortcut like <kbd>ctrl</kbd> + <kbd>&lt;i&gt;</kbd>. That would match many other programs behaviour. For example browsers, some terminals (Windows terminal uses <kbd>ctrl</kbd>+<kbd>alt</kbd> + <kdb>&lt;i&gt;</kbd>, and iTerm 2 on MacOS <kbd>cmd</kbd>+<kbd>&lt;i&gt;</kbd>), Windows explorer, and probably a lot more I can't think of right now. Interestingly, VS Code choose <kbd>ctrl</kbd> + <kbd>&lt;i&gt;</kbd> as shortcut for pane (you call those "Windows") switching, and <kbd>alt</kbd> + <kbd>&lt;i&gt;</kbd> for tabs. That's probably why I've never used it there. Instead, I mainly utilized <kbd>ctrl</kbd> + <kbd>tab</kbd> (it cycles through the open editors).

I would be happy to implement more functionality, but I'm not really sure if that would collide with other shortcuts, nor have I figured out how to do that yet :D.

**Comment by Elias-Graf** [comment]
## Update: Added shortcut support 

I've now added support for shortcuts in the previously mentioned pull request. I've gone with the combinations that I was already most familiar with (see above).
One can now do: <kbd>ctrl</kbd>+<kbd>&lt;[1-9]&gt;</kbd> to quickly access buffers. I've already found it very nice to use, and even tried to use it while editing my fork with the upstream version, which sadly didn't work, lol.

**Comment by DreckSallow** [comment]
Hello! I've added a PR #8198  to close this issue because it has been open for a long time.
